server/server.go

```diff
@@ -0,0 +1,130 @@
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	"golang.org/x/oauth2"
+	"google.golang.org/api/option"
+	secretmanagerpb "google.golang.org/genproto/googleapis/cloud/secretmanager/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"sigs.k8s.io/secrets-store-csi-driver/provider/v1alpha1"
+)
+
+type Server struct {
+	UA             string
+	RuntimeVersion string
+	Kubeconfig     string // TODO: accept a kubernetes.Clientset instead
+}
+
+var _ v1alpha1.CSIDriverProviderServer = &Server{}
+
+// Mount implements provider csi-provider method
+func (s *Server) Mount(ctx context.Context, req *v1alpha1.MountRequest) (*v1alpha1.MountResponse, error) {
+	p, err := strconv.ParseUint(req.GetPermission(), 10, 32)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("Unable to parse permissions: %s", req.GetPermission()))
+
+	}
+
+	params := &config.MountParams{
+		Attributes:  req.GetAttributes(),
+		KubeSecrets: req.GetSecrets(),
+		TargetPath:  req.GetTargetPath(),
+		Permissions: os.FileMode(p),
+	}
+
+	cfg, err := config.Parse(params)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	smOpts := []option.ClientOption{option.WithUserAgent(s.UA)}
+
+	// Build the workload identity auth token if possible (fallback to node ID)
+	token, err := auth.Token(ctx, cfg, s.Kubeconfig)
+	if err != nil {
+		log.Printf("unable to use workload identity: %v", err)
+	} else {
+		smOpts = append(smOpts, option.WithTokenSource(oauth2.StaticTokenSource(token)))
+	}
+
+	// Build the secret manager client
+	client, err := secretmanager.NewClient(ctx, smOpts...)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to create secretmanager client: %v", err))
+	}
+
+	// Fetch the secrets from the secretmanager API and write them to the
+	// filesystem based on the SecretProviderClass configuration.
+	ovs, err := handleMountEvent(ctx, client, cfg)
+
+	return &v1alpha1.MountResponse{
+		ObjectVersion: ovs,
+	}, err
+}
+
+// Version implements provider csi-provider method
+func (s *Server) Version(ctx context.Context, req *v1alpha1.VersionRequest) (*v1alpha1.VersionResponse, error) {
+	return &v1alpha1.VersionResponse{
+		Version:        "v1alpha1",
+		RuntimeName:    "secrets-store-csi-driver-provider-gcp",
+		RuntimeVersion: s.RuntimeVersion,
+	}, nil
+}
+
+// handleMountEvent fetches the secrets from the secretmanager API and
+// writes them to the filesystem based on the SecretProviderClass configuration.
+func handleMountEvent(ctx context.Context, client *secretmanager.Client, cfg *config.MountConfig) ([]*v1alpha1.ObjectVersion, error) {
+	ovs := []*v1alpha1.ObjectVersion{}
+	for _, secret := range cfg.Secrets {
+		req := &secretmanagerpb.AccessSecretVersionRequest{
+			Name: secret.ResourceName,
+		}
+
+		result, err := client.AccessSecretVersion(ctx, req)
+		if err != nil {
```

Comment:
If there is an error, do we want it to be all or nothing instead of partial failure? If we want partial failure, should we continue to attempt on the rest of the secrets?

Reply:
Filed https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp/issues/52 for this, I can see partial failures being good but we may need to sync with the driver project on other provider behaviors.